### PR TITLE
bpo-42064: Pass module state to trace, progress, and authorizer callbacks

### PR DESCRIPTION
**Pull request by erlend-aasland**
- add print-or-clear traceback helper
- add helpers to clear and visit saved contexts
- modify callbacks to use the new callback_context struct

https://bugs.python.org/issue42064

### CODE REVIEW

**Comment by erlend-aasland** [comment]
On hold until #27934 is merged. There should be no conflicts with #27931.

File: Modules/_sqlite/connection.c
Comment (encukou) [inline Modules/_sqlite/connection.c]: These three statements are named `free_callback_contexts()` below. But here they are problematic.
* If this is a brand new connection, then `PyType_GenericAlloc` has set all these to NULL and they don't need to be NULLed again.
* If someone is calling `__init__` for the second time, this might free the Python callable, but SQLite will still have the pointer. Also the old `db` will leak, and who knows what'll happen to old cursors… Whoa, I don't see how reinitialization could work; should it rather raise an exception?

File: Modules/_sqlite/connection.c
Comment (encukou) [inline Modules/_sqlite/connection.c]: Suggestion: Be defensive and initialize this to `SQLITE_DENY`?

**Review by encukou — COMMENTED** [review]
Nitpick:
`SET_CALLBACK_CONTEXT` and `CLEAR_CALLBACK_CONTEXT` can both be regular functions rather than macros, making them shorter and more readable.
That leaves `VISIT_CALLBACK_CONTEXT`, which I think is too robust for the 3 uses. But that's very much a personal opinion.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I had reinitialisation in mind, but as you say, it's hard to imagine how to do that properly anyway. Maybe implicit connection close at the start of `__init__` will do the trick. I'll have to think that through. Perhaps we should fix the reinit case before proceeding with this PR.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I agree.

File: Modules/_sqlite/connection.c
Comment (encukou) [reply Modules/_sqlite/connection.c]: > Maybe implicit connection close at the start of `__init__` will do the trick.

I would be surprised. Cursors seem to have hard (and reasonable!) assumptions about their connection's `db`.
I also don't see the use case – when would a new connection object not be enough?

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Hm, true. ~Let's raise an exception instead.~

I did some tests, and reinit actually seems to work pretty well using an implicit close in `__init__`, even with dangling cursors that fetch and execute after reinit. There's no refleaks on our side, and SQLite seems to be able to clean up as well. The only improvement would be to clear the callbacks from SQLite before the implicit close.

> I also don't see the use case

Me neither.

**Comment by erlend-aasland** [comment]
> SET_CALLBACK_CONTEXT and CLEAR_CALLBACK_CONTEXT can both be regular functions rather than macros, making them shorter and more readable.
> That leaves VISIT_CALLBACK_CONTEXT, which I think is too robust for the 3 uses. But that's very much a personal opinion.

- `VISIT_CALLBACK_CONTEXT` cannot be implemented as a regular function since it uses `Py_VISIT`; I prefer to keep the macro
- I'll change `CLEAR_CALLBACK_CONTEXT` to a regular function; I have no strong opinion about this :)
- `SET_CALLBACK_CONTEXT` will IMO be more complex as a function, since it would take a pointer to a pointer as an argument; if you have a strong opinion about this, I'll of course change it

**Comment by erlend-aasland** [comment]
NB, this needs to be rebased onto `main` after #28088 is merged, assuming of course that it'll be easier to land that PR.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c:973]: I just stumbled over a bug in the trace callback:
The trace callback checks `enable_callback_tracebacks` and either `PyErr_Print` (and clear) the current exception, or `PyErr_Clear` the current exception. That's fine for the function, step, and final callbacks; they all pass an error back to SQLite, and we translate that into an exception after checking the return value of `sqlite3_step()` or `sqlite3_finalize()`. However, for the trace callback, we _always_ clear the current exception:

```
Python 3.10.0rc1 (v3.10.0rc1:cc115e5bf6, Aug  2 2021, 17:54:14) [Clang 12.0.5 (clang-1205.0.22.11)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(lambda x: 5/0)
>>> try:
...     cx.execute("select 1")
... except:
...     print("exception occurred")
... else:
...     print("no exception occurred")
... 
<sqlite3.Cursor object at 0x1006dc640>
no exception occurred
>>> sqlite3.enable_callback_tracebacks(True)
>>> try: 
...     cx.execute("select 1")
... except:
...     print("exception occurred")
... else:
...     print("no exception occurred")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <lambda>
ZeroDivisionError: division by zero
<sqlite3.Cursor object at 0x100891c40>
no exception occurred
```

I cannot imagine that this was the intended behaviour.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:973]: I opened #28133 for this.

File: Modules/_sqlite/connection.c
Comment (encukou) [reply Modules/_sqlite/connection.c]: >  There's no refleaks on our side, and SQLite seems to be able to clean up as well. 

I'm surprised. Wouldn't this result in a `sqlite3_open_v2` without a corresponding close?

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Perhaps I was a little bit vague. I did the tests with an added implicit close in `__init__`.

Here's the diff I used:
```diff
diff --git a/Modules/_sqlite/connection.c b/Modules/_sqlite/connection.c
index 63fd916662..961afe8de0 100644
--- a/Modules/_sqlite/connection.c
+++ b/Modules/_sqlite/connection.c
@@ -56,6 +56,7 @@ static const char * const begin_statements[] = {
 static int pysqlite_connection_set_isolation_level(pysqlite_Connection* self, PyObject* isolation_level, void *Py_UNUSED(ignored));
 static void _pysqlite_drop_unused_cursor_references(pysqlite_Connection* self);
 static void free_callback_context(callback_context *ctx);
+static void connection_close(pysqlite_Connection *self);
 
 static PyObject *
 new_statement_cache(pysqlite_Connection *self, int maxsize)
@@ -114,6 +115,18 @@ pysqlite_connection_init_impl(pysqlite_Connection *self,
         return -1;
     }
 
+    // Gracefully handle reinitialization
+    if (self->initialized) {
+        (void)sqlite3_set_authorizer(self->db, NULL, NULL);
+        sqlite3_progress_handler(self->db, 0, NULL, NULL);
+#ifdef HAVE_TRACE_V2
+        sqlite3_trace_v2(self->db, SQLITE_TRACE_STMT, NULL, NULL);
+#else
+        sqlite3_trace(self->db, NULL, NULL);
+#endif
+        connection_close(self);
+    }
+
     pysqlite_state *state = pysqlite_get_state_by_type(Py_TYPE(self));
     self->state = state;
 


```

Comment (encukou) [reply Modules/_sqlite/connection.c]: I see. Well, this is more complicated than I initially thought; I guess it should be a separate bpo issue.

I keep finding issues:

- `db` is not set to NULL if init fails.
- This segfaults for me:

```python
import sqlite3

conn = sqlite3.connect(':memory:')
conn.execute('CREATE TABLE foo (bar)')

try:
    conn.__init__('/bad-file/')
except sqlite3.OperationalError:
    pass

conn.execute('INSERT INTO foo (bar) VALUES (1), (2), (3), (4)')
```
... and even if that's fixed, I'm concerned that there will be other edge cases we can't reasonably think about.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Yes, there's some issues in connection init. I'd gladly clean it up and harden it before continuing with this PR.

**Comment by encukou** [comment]
> will IMO be more complex as a function, since it would take a pointer to a pointer as an argument

Personally I'd prefer a type-safe function with a more complex argument, but as I said, this is nitpicking. Feel free to make them all macros for consistency, if you find it easier to read :)

I plan to review #28088 first.

**Comment by erlend-aasland** [comment]
> > will IMO be more complex as a function, since it would take a pointer to a pointer as an argument
> 
> Personally I'd prefer a type-safe function with a more complex argument, but as I said, this is nitpicking. Feel free to make them all macros for consistency, if you find it easier to read :)

I'll keep `VISIT_CALLBACK_CONTEXT` function a macro in any case, but I'll think thrice about `SET_CALLBACK_CONTEXT` before deciding :)

> I plan to review #28088 first.

Great, thanks!

BTW, I'm creating a bpo for cleaning up connection `__init__`; I've got a WIP branch lying around already, so you can expect a PR anytime soon :)

File: Modules/_sqlite/connection.c
Comment (encukou) [reply Modules/_sqlite/connection.c]: It's not a blocker for this PR; just please open a bpo so the issues won't be lost in a closed PR.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I've opened issue 45126 for this.

**Comment by erlend-aasland** [comment]
All right, no more changes comin' up, unless you've got further remarks :)

**Review by encukou — APPROVED** [review]
Looks good, thanks!

**Comment by erlend-aasland** [comment]
> Looks good, thanks!

Likewise!

**Pull request merged**
